### ISSUE TITLE
Enhancement: Complete drop-in replacement

### ISSUE THREAD

**Issue by UInt2048** [UInt2048/CyberKit#30]
As of v0.0.8 alpha, CyberKit was released with renamed frameworks, such as:

* CyberScriptCore.framework (replaces JavaScriptCore)
* CyberCore.framework (replaces WebCore)
* CyberKitLegacy.framework (replaces WebKitLegacy)
* CyberKit.framework (replaces WebKit)
* libwebrtc.dylib (replaces libwebrtc)
* libANGLE-shared.dylib (replaces libANGLE-shared)
* libswiftCyberKit.dylib (replaces libswiftWebKit)

Due to the renaming, the symbols in libswiftCyberKit are not a drop-in replacement for WebKit's (since the symbols depend on the name of CyberKit.framework / WebKit.framework).

CyberKit intended to end the renaming, but was met with a crash log like the following.

A null pointer dereference such as shown here in the CoreFoundation method `_CFGetNonObjCTypeID` is the result of a bundle identifier mismatch between the bundle identifier used in the CFBundleRef (typically created via `CFBundleGetBundleWithIdentifier`) and the actual value of the bundle identifier.

In upstream WebKit and in CyberKit, target bundle identifiers are set via configuration files, which should be revisited in the event of this type of crash.

<details>
<summary> Crash Log (with translation) </summary>

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Incident Identifier: 81F7D3B0-2E53-4273-934D-79DDB432AF2B
CrashReporter Key:   9d66e1f8543fafd1530513f05915f393302a1abf
Hardware Model:      iPhone15,2
Process:             com.matthewbenedict.WebKit.WebContent [15068]
Path:                /private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/XPCServices/com.matthewbenedict.WebKit.WebContent.xpc/com.matthewbenedict.WebKit.WebContent
Identifier:          com.matthewbenedict.WebKit.WebContent
Version:             617+ (617.1.4.3+)
Code Type:           ARM-64 (Native)
Role:                Foreground
Parent Process:      launchd [1]
Coalition:           com.matthewbenedict.MobileMiniBrowser [1833]

Date/Time:           2024-01-28 05:28:59.7269 -1100
Launch Time:         2024-01-28 05:28:59.2334 -1100
OS Version:          iPhone OS 16.2 (20C65)
Release Type:        User
Baseband Version:    1.22.02
Report Version:      104

Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x0000000000000008
Exception Codes: 0x0000000000000001, 0x0000000000000008
VM Region Info: 0x8 is not in any region.  Bytes before following region: 68719476728
      REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      commpage (reserved)     1000000000-7000000000 [384.0G] ---/--- SM=NUL  ...(unallocated)
Termination Reason: SIGNAL 11 Segmentation fault: 11
Terminating Process: exc handler [15068]

Triggered by Thread:  1

Thread 0 name:   Dispatch queue: com.apple.main-thread
Thread 0:
0   libsystem_kernel.dylib        	       0x1cbcc8aa8 mach_msg2_trap + 8
1   libsystem_kernel.dylib        	       0x1cbcdafc4 mach_msg2_internal + 79
2   libsystem_kernel.dylib        	       0x1cbcdb204 mach_msg_overwrite + 387
3   libsystem_kernel.dylib        	       0x1cbcc8fec mach_msg + 23
4   CoreFoundation                	       0x18e3cead4 __CFRunLoopServiceMachPort + 159
5   CoreFoundation                	       0x18e3cfd18 __CFRunLoopRun + 1231
6   CoreFoundation                	       0x18e3d4ec0 CFRunLoopRunSpecific + 611
7   Foundation                    	       0x1887130d4 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 211
8   Foundation                    	       0x188712f68 -[NSRunLoop(NSRunLoop) run] + 63
9   libxpc.dylib                  	       0x1dc4433a8 _xpc_objc_main + 495
10  libxpc.dylib                  	       0x1dc4456a0 xpc_main + 155
11  WebKit                        	       0x10f6bcd10 WebKit::XPCServiceMain(int, char const**) + 172
12  WebKit                        	       0x1116cc600 WKXPCServiceMain + 32
13  ...wbenedict.WebKit.WebContent	       0x1003ac3b8 main + 220
14  dyld                          	       0x1acbf6960 start + 2527

Thread 1 name:   Dispatch queue: com.apple.root.default-qos
Thread 1 Crashed:
0   CoreFoundation                	       0x18e3d5954 _CFGetNonObjCTypeID + 92
1   CoreFoundation                	       0x18e443448 CFBundleGetFunctionPointerForName + 27
2   WebKit                        	       0x10f6cda04 invocation function for block in WebKit::XPCServiceEventHandler(NSObject<OS_xpc_object>*) + 1176
3   libxpc.dylib                  	       0x1dc43a79c _xpc_connection_call_event_handler + 151
4   libxpc.dylib                  	       0x1dc43abcc _xpc_connection_mach_event + 991
5   libdispatch.dylib             	       0x19597109c _dispatch_client_callout4 + 19
6   libdispatch.dylib             	       0x19598d830 _dispatch_mach_msg_invoke + 467
7   libdispatch.dylib             	       0x19597856c _dispatch_lane_serial_drain + 375
8   libdispatch.dylib             	       0x19598e54c _dispatch_mach_invoke + 447
9   libdispatch.dylib             	       0x195983e10 _dispatch_workloop_worker_thread + 651
10  libsystem_pthread.dylib       	       0x1dc3e5df8 _pthread_wqthread + 287
11  libsystem_pthread.dylib       	       0x1dc3e5b98 start_wqthread + 7

Thread 2:
0   libsystem_pthread.dylib       	       0x1dc3e5b90 start_wqthread + 0


Thread 1 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000000000   x1: 0x0000000112939aa8   x2: 0x0000000000001603   x3: 0x000000015fd0d140
    x4: 0x0000000000000060   x5: 0x000000015fd0cc20   x6: 0x0000000000000058   x7: 0x0000000000000000
    x8: 0x0000000000000008   x9: 0x211c272d61505333  x10: 0x0000000000000000  x11: 0x000000018e594e34
   x12: 0x0000000111ae572b  x13: 0x0000000000000074  x14: 0x0000000111ae5729  x15: 0x0000000111ae5729
   x16: 0x000000018e44342c  x17: 0x00000000586c321e  x18: 0x0000000000000000  x19: 0x0000000112939aa8
   x20: 0x0000000000000000  x21: 0x000000015fd0be90  x22: 0x0000000000000000  x23: 0x0000000000010000
   x24: 0x0000000000000000  x25: 0x000000016fadf0e0  x26: 0x000000016fade680  x27: 0x000000015fd05ff8
   x28: 0x0000000161006e20   fp: 0x000000016fade260   lr: 0xd51f00018e443448
    sp: 0x000000016fade260   pc: 0x000000018e3d5954 cpsr: 0x60001000
   far: 0x0000000000000008  esr: 0x92000006 (Data Abort) byte read Translation fault

Binary Images:
       0x1cbcc4000 -        0x1cbcfefe3 libsystem_kernel.dylib arm64e  <9daa5c2993e03768a3e1e139995dc4af> /usr/lib/system/libsystem_kernel.dylib
       0x18e354000 -        0x18e739fff CoreFoundation arm64e  <725e49f4653b39bf9a7a8a3250911ecb> /System/Library/Frameworks/CoreFoundation.framework/CoreFoundation
       0x1886d1000 -        0x18901afff Foundation arm64e  <07a92f05d8ec327eab3341db9f77ba16> /System/Library/Frameworks/Foundation.framework/Foundation
       0x1dc429000 -        0x1dc469fff libxpc.dylib arm64e  <9826b3aa0ca03ca4b664d86869fb33f2> /usr/lib/system/libxpc.dylib
       0x10e8c0000 -        0x112897fff WebKit arm64  <32d4a06f12cb3b4b9cf9a1df8e6ef813> /private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/WebKit
       0x1003a8000 -        0x1003affff com.matthewbenedict.WebKit.WebContent arm64  <03f497394b1131fa8cc9a65b29cdd3a3> /private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/XPCServices/com.matthewbenedict.WebKit.WebContent.xpc/com.matthewbenedict.WebKit.WebContent
       0x1acbe1000 -        0x1acc6408f dyld arm64e  <7b63c57361613b33a3a29944ba59722f> /usr/lib/dyld
       0x19596d000 -        0x1959b3fff libdispatch.dylib arm64e  <24db930d870b31ceac69534d7896b4a8> /usr/lib/system/libdispatch.dylib
       0x1dc3e5000 -        0x1dc3f0fff libsystem_pthread.dylib arm64e  <f2ba7ec0f75a3345b4f6f7da4979b902> /usr/lib/system/libsystem_pthread.dylib
               0x0 - 0xffffffffffffffff ??? unknown-arch  <00000000000000000000000000000000> ???

EOF

-----------
Full Report
-----------

{"app_name":"com.matthewbenedict.WebKit.WebContent","timestamp":"2024-01-28 05:29:03.00 -1100","app_version":"617+","slice_uuid":"03f49739-4b11-31fa-8cc9-a65b29cdd3a3","build_version":"617.1.4.3+","platform":"2","bundleID":"com.matthewbenedict.WebKit.WebContent","share_with_app_devs":"0","is_first_party":"1","bug_type":"309","os_version":"iPhone OS 16.2 (20C65)","roots_installed":"0","name":"com.matthewbenedict.WebKit.WebContent","incident_id":"81F7D3B0-2E53-4273-934D-79DDB432AF2B"}
{
"uptime": 610000,
"procRole": "Foreground",
"version": 2,
"userID": 501,
"deployVersion": 210,
"modelCode": "iPhone15,2",
"coalitionID": 1833,
"osVersion": {
"isEmbedded": true,
"train": "iPhone OS 16.2",
"releaseType": "User",
"build": "20C65"
},
"captureTime": "2024-01-28 05:28:59.7269 -1100",
"incident": "81F7D3B0-2E53-4273-934D-79DDB432AF2B",
"pid": 15068,
"cpuType": "ARM-64",
"roots_installed": 0,
"bug_type": "309",
"procLaunch": "2024-01-28 05:28:59.2334 -1100",
"procStartAbsTime": 14683516746502,
"procExitAbsTime": 14683528575609,
"procName": "com.matthewbenedict.WebKit.WebContent",
"procPath": "/private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/XPCServices/com.matthewbenedict.WebKit.WebContent.xpc/com.matthewbenedict.WebKit.WebContent",
"bundleInfo": {
"CFBundleShortVersionString": "617+",
"CFBundleVersion": "617.1.4.3+",
"CFBundleIdentifier": "com.matthewbenedict.WebKit.WebContent"
},
"parentProc": "launchd",
"parentPid": 1,
"coalitionName": "com.matthewbenedict.MobileMiniBrowser",
"crashReporterKey": "9d66e1f8543fafd1530513f05915f393302a1abf",
"wasUnlockedSinceBoot": 1,
"isLocked": 0,
"throttleTimeout": 10,
"basebandVersion": "1.22.02",
"vmRegionInfo": "0x8 is not in any region.  Bytes before following region: 68719476728\n      REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      commpage (reserved)     1000000000-7000000000 [384.0G] ---/--- SM=NUL  ...(unallocated)",
"exception": {
"codes": "0x0000000000000001, 0x0000000000000008",
"rawCodes": [
1,
8
],
"type": "EXC_BAD_ACCESS",
"signal": "SIGSEGV",
"subtype": "KERN_INVALID_ADDRESS at 0x0000000000000008"
},
"termination": {
"flags": 0,
"code": 11,
"namespace": "SIGNAL",
"indicator": "Segmentation fault: 11",
"byProc": "exc handler",
"byPid": 15068
},
"vmregioninfo": "0x8 is not in any region.  Bytes before following region: 68719476728\n      REGION TYPE                 START - END      [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      commpage (reserved)     1000000000-7000000000 [384.0G] ---/--- SM=NUL  ...(unallocated)",
"faultingThread": 1,
"threads": [
{
"id": 6006068,
"queue": "com.apple.main-thread",
"frames": [
{
"imageOffset": 19112,
"symbol": "mach_msg2_trap",
"symbolLocation": 8,
"imageIndex": 0
},
{
"imageOffset": 94148,
"symbol": "mach_msg2_internal",
"symbolLocation": 79,
"imageIndex": 0
},
{
"imageOffset": 94724,
"symbol": "mach_msg_overwrite",
"symbolLocation": 387,
"imageIndex": 0
},
{
"imageOffset": 20460,
"symbol": "mach_msg",
"symbolLocation": 23,
"imageIndex": 0
},
{
"imageOffset": 502484,
"symbol": "__CFRunLoopServiceMachPort",
"symbolLocation": 159,
"imageIndex": 1
},
{
"imageOffset": 507160,
"symbol": "__CFRunLoopRun",
"symbolLocation": 1231,
"imageIndex": 1
},
{
"imageOffset": 528064,
"symbol": "CFRunLoopRunSpecific",
"symbolLocation": 611,
"imageIndex": 1
},
{
"imageOffset": 270548,
"symbol": "-[NSRunLoop(NSRunLoop) runMode:beforeDate:]",
"symbolLocation": 211,
"imageIndex": 2
},
{
"imageOffset": 270184,
"symbol": "-[NSRunLoop(NSRunLoop) run]",
"symbolLocation": 63,
"imageIndex": 2
},
{
"imageOffset": 107432,
"symbol": "_xpc_objc_main",
"symbolLocation": 495,
"imageIndex": 3
},
{
"imageOffset": 116384,
"symbol": "xpc_main",
"symbolLocation": 155,
"imageIndex": 3
},
{
"imageOffset": 14667024,
"symbol": "WebKit::XPCServiceMain(int, char const**)",
"symbolLocation": 172,
"imageIndex": 4
},
{
"imageOffset": 48285184,
"symbol": "WKXPCServiceMain",
"symbolLocation": 32,
"imageIndex": 4
},
{
"imageOffset": 17336,
"symbol": "main",
"symbolLocation": 220,
"imageIndex": 5
},
{
"imageOffset": 88416,
"symbol": "start",
"symbolLocation": 2527,
"imageIndex": 6
}
]
},
{
"triggered": true,
"id": 6006076,
"threadState": {
"x": [
{
"value": 0
},
{
"value": 4606630568
},
{
"value": 5635
},
{
"value": 5902487872
},
{
"value": 96
},
{
"value": 5902486560
},
{
"value": 88
},
{
"value": 0
},
{
"value": 8
},
{
"value": 2385824978459185971
},
{
"value": 0
},
{
"value": 6683184692,
"symbolLocation": 0,
"symbol": "__CFMacRomanCharToUnicharTable"
},
{
"value": 4591605547
},
{
"value": 116
},
{
"value": 4591605545
},
{
"value": 4591605545
},
{
"value": 6681801772,
"symbolLocation": 0,
"symbol": "CFBundleGetFunctionPointerForName"
},
{
"value": 1483485726
},
{
"value": 0
},
{
"value": 4606630568
},
{
"value": 0
},
{
"value": 5902483088
},
{
"value": 0
},
{
"value": 65536
},
{
"value": 0
},
{
"value": 6168637664
},
{
"value": 6168635008
},
{
"value": 5902458872
},
{
"value": 5922385440
}
],
"flavor": "ARM_THREAD_STATE64",
"lr": {
"value": 15356993261038482504
},
"cpsr": {
"value": 1610616832
},
"fp": {
"value": 6168633952
},
"sp": {
"value": 6168633952
},
"esr": {
"value": 2449473542,
"description": "(Data Abort) byte read Translation fault"
},
"pc": {
"value": 6681352532,
"matchesCrashFrame": 1
},
"far": {
"value": 8
}
},
"queue": "com.apple.root.default-qos",
"frames": [
{
"imageOffset": 530772,
"symbol": "_CFGetNonObjCTypeID",
"symbolLocation": 92,
"imageIndex": 1
},
{
"imageOffset": 980040,
"symbol": "CFBundleGetFunctionPointerForName",
"symbolLocation": 27,
"imageIndex": 1
},
{
"imageOffset": 14735876,
"symbol": "invocation function for block in WebKit::XPCServiceEventHandler(NSObject<OS_xpc_object>*)",
"symbolLocation": 1176,
"imageIndex": 4
},
{
"imageOffset": 71580,
"symbol": "_xpc_connection_call_event_handler",
"symbolLocation": 151,
"imageIndex": 3
},
{
"imageOffset": 72652,
"symbol": "_xpc_connection_mach_event",
"symbolLocation": 991,
"imageIndex": 3
},
{
"imageOffset": 16540,
"symbol": "_dispatch_client_callout4",
"symbolLocation": 19,
"imageIndex": 7
},
{
"imageOffset": 133168,
"symbol": "_dispatch_mach_msg_invoke",
"symbolLocation": 467,
"imageIndex": 7
},
{
"imageOffset": 46444,
"symbol": "_dispatch_lane_serial_drain",
"symbolLocation": 375,
"imageIndex": 7
},
{
"imageOffset": 136524,
"symbol": "_dispatch_mach_invoke",
"symbolLocation": 447,
"imageIndex": 7
},
{
"imageOffset": 93712,
"symbol": "_dispatch_workloop_worker_thread",
"symbolLocation": 651,
"imageIndex": 7
},
{
"imageOffset": 3576,
"symbol": "_pthread_wqthread",
"symbolLocation": 287,
"imageIndex": 8
},
{
"imageOffset": 2968,
"symbol": "start_wqthread",
"symbolLocation": 7,
"imageIndex": 8
}
]
},
{
"id": 6006077,
"frames": [
{
"imageOffset": 2960,
"symbol": "start_wqthread",
"symbolLocation": 0,
"imageIndex": 8
}
]
}
],
"usedImages": [
{
"source": "P",
"arch": "arm64e",
"base": 7714127872,
"size": 241636,
"uuid": "9daa5c29-93e0-3768-a3e1-e139995dc4af",
"path": "/usr/lib/system/libsystem_kernel.dylib",
"name": "libsystem_kernel.dylib"
},
{
"source": "P",
"arch": "arm64e",
"base": 6680821760,
"size": 4087808,
"uuid": "725e49f4-653b-39bf-9a7a-8a3250911ecb",
"path": "/System/Library/Frameworks/CoreFoundation.framework/CoreFoundation",
"name": "CoreFoundation"
},
{
"source": "P",
"arch": "arm64e",
"base": 6583816192,
"size": 9740288,
"uuid": "07a92f05-d8ec-327e-ab33-41db9f77ba16",
"path": "/System/Library/Frameworks/Foundation.framework/Foundation",
"name": "Foundation"
},
{
"source": "P",
"arch": "arm64e",
"base": 7990317056,
"size": 266240,
"uuid": "9826b3aa-0ca0-3ca4-b664-d86869fb33f2",
"path": "/usr/lib/system/libxpc.dylib",
"name": "libxpc.dylib"
},
{
"source": "P",
"arch": "arm64",
"base": 4539023360,
"size": 66945024,
"uuid": "32d4a06f-12cb-3b4b-9cf9-a1df8e6ef813",
"path": "/private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/WebKit",
"name": "WebKit"
},
{
"source": "P",
"arch": "arm64",
"base": 4298801152,
"size": 32768,
"uuid": "03f49739-4b11-31fa-8cc9-a65b29cdd3a3",
"path": "/private/var/containers/Bundle/Application/A5E73BD3-44BE-42FD-A7D2-C13B1E53E275/MobileMiniBrowser.app/Frameworks/WebKit.framework/XPCServices/com.matthewbenedict.WebKit.WebContent.xpc/com.matthewbenedict.WebKit.WebContent",
"name": "com.matthewbenedict.WebKit.WebContent"
},
{
"source": "P",
"arch": "arm64e",
"base": 7193104384,
"size": 536720,
"uuid": "7b63c573-6161-3b33-a3a2-9944ba59722f",
"path": "/usr/lib/dyld",
"name": "dyld"
},
{
"source": "P",
"arch": "arm64e",
"base": 6804656128,
"size": 290816,
"uuid": "24db930d-870b-31ce-ac69-534d7896b4a8",
"path": "/usr/lib/system/libdispatch.dylib",
"name": "libdispatch.dylib"
},
{
"source": "P",
"arch": "arm64e",
"base": 7990038528,
"size": 49152,
"uuid": "f2ba7ec0-f75a-3345-b4f6-f7da4979b902",
"path": "/usr/lib/system/libsystem_pthread.dylib",
"name": "libsystem_pthread.dylib"
},
{
"size": 0,
"source": "A",
"base": 0,
"uuid": "00000000-0000-0000-0000-000000000000"
}
],
"sharedCache": {
"base": 6565494784,
"size": 2944221184,
"uuid": "2b5dcb8b-649b-3c0a-bb4b-98268b3fd189"
},
"vmSummary": "ReadOnly portion of Libraries: Total=1.5G resident=0K(0%) swapped_out_or_unallocated=1.5G(100%)\nWritable regions: Total=63.3M written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=63.3M(100%)\n\n                                VIRTUAL   REGION \nREGION TYPE                        SIZE    COUNT (non-coalesced) \n===========                     =======  ======= \nActivity Tracing                   256K        1 \nDispatch continuations            6144K        1 \nKernel Alloc Once                   32K        1 \nMALLOC                            54.2M       20 \nMALLOC guard page                   64K        4 \nSTACK GUARD                         48K        3 \nStack                             2096K        3 \n__AUTH                            1660K      242 \n__AUTH_CONST                      16.3M      437 \n__CTF                               756        1 \n__DATA                            6211K      436 \n__DATA_CONST                      23.8M      449 \n__DATA_DIRTY                      1637K      371 \n__FONT_DATA                        2352        1 \n__LINKEDIT                       820.0M        9 \n__OBJC_CONST                      3535K      266 \n__OBJC_RO                         78.8M        1 \n__OBJC_RW                         2436K        1 \n__TEXT                           703.9M      456 \ndyld private memory                256K        1 \nmapped file                         64K        1 \nshared memory                       32K        2 \n===========                     =======  ======= \nTOTAL                              1.7G     2707 \n",
"legacyInfo": {
"threadTriggered": {
"queue": "com.apple.root.default-qos"
}
},
"trialInfo": {
"rollouts": [
{
"rolloutId": "62b4513af75dc926494899c6",
"factorPackIds": {
"COREOS_ICD": "62b451816c843c3494e8dd8f"
},
"deploymentId": 250000012
},
{
"rolloutId": "6112d17137f5d11121dcd4e2",
"factorPackIds": {
"SIRI_UNDERSTANDING_ASR_HAMMER": "638a086875b0720ff1f8c2bb"
},
"deploymentId": 250000180
}
],
"experiments": []
}
}

```

</details>

**Comment by UInt2048** [comment]
9ac4c44044f24226f5bdb80d45516ee0348cdc80 might be useful for debugging purposes

**Comment by UInt2048** [comment]
Update: This issue is rename-independent. Renaming the binaries to CyberKit equivalents does NOT fix the issue at 01b7a82

**Comment by ghh-jb** [comment]
I think it cant be that "drop-in replacement" for now, because we are unable just to bindmount it on top of /System/Library/Frameworks/WebKit.framework and others...

If only we could do that - we could fix ALL issues with web pages on all jailbreak-supported devices!
That will pretty easy to use but it is not that hard to implement i think...
@UInt2048
